Add otel tracing

cpuguy83 opened the pull request:
- Adds basic otel instrumentation to http transports/handlers
- Adds OTLP/HTTP exporter
- Updates tests to configure traces/spans automatically (as automatically as we can, anyway)
- Exposes options in Makefile to pass through info needed to export traces from test daemon

thaJeztah commented:
We had a (lengthy) discussion about this PR in Yesterday's sync, looking at:

- How OTEL / telemetry is commonly used
- Ways to configure (daemon.json, daemon flags, environment variables)
- Configuring on Windows systems
- Defaults (enabled by default, disabled by default)

I tried to capture the discussion below.

### How OTEL / telemetry is commonly used

While this PR (obviously) is implementing OTEL support in the Moby Engine, the Engine is just "one" component in the container stack (the `docker` CLI, BuildKit (integrated), containerd, runc, etc). When using OTEL, a common scenario would be that the user wants to collect telemetry of all components of the stack (and, not unlikely, other services running on the same host). In suchb a scenario, all of these components should be configured equally, so that telemetry of all components can be aggregated.

### Ways to configure (daemon.json, daemon flags, environment variables)

We generally recommend configuring the Engine through it's configuration file (`daemon.json`), but provide command-line flags for scenarios where alternatives are more appropriate (manually starting the daemon from the commandline for debugging, or setting options through a systemd unit). Allowing the user to configure options through the `daemon.json` configuration file allows the user to keep all options in a single location, as well as allowing us to explicitly document available configuration options.

From this perspective, we discussed adding configuration options on the daemon to configure OTEL options. This posed some challenges;

- As outlined in "How OTEL / telemetry is commonly used", setting up opentelemetry commonly involves configuring it _globally_ (i.e. configuring all tools that support OTEL with the same options).
- The standard way to configure these options is through environment variables that are defined by the OTEL libraries; doing so allows for environment variables to be configured on the system, and consumed by all tools providing OTEL.
- OTEL is still quite actively developed; new features (and configuration options), which means that it's currently difficult to define a hard-coded list of options that can be configured, so we'd either need to define a generic (`map[string]any`) set of options, or play "catch up" with new features added.
- Because of the above, configuring OTEL options through daemon flags or the `daemon.json` configuration file would make setting up OTEL for a host _more complicated_, as configuration options would now be set in multiple locations (environment variables for some tools, configuration file(s) for the Moby Engine).

Based on the above, we concluded that using environment variables (as defined by the OTEL libraries) would be the most logical way to configure OTEL options.

### Configuring on Windows systems

Setting up environment for Windows services is known to be "more involved" than Linux. We discussed if we needed to provide alternative options (as we did for setting up proxy-configuration through daemon.json (https://github.com/moby/moby/pull/42835)). An option we discussed was to provide a generic "env-vars" (key/value: `map[string]string`) option in the `daemon.json`. This option would allow setting arbitrary environment variables. The daemon would read these options, and `os.SetEnv()` these during startup (potentially during reload), and if needed we could implement an allow-list of environment variables that can be set.

```json
{
  "env-vars": {
    "OTEL_TRACES_EXPORTER": "otlp",
    "OTEL_EXPORTER_OTLP_ENDPOINT": "...",
    "OTEL_EXPORTER_OTLP_TRACES_ENDPOINT": "..."
  }
}
```

While this could be a future enhancement if there's a need, we decided to _not_ implement such an option for now. As mentioned before, it's likely that other tools also would need to be configured for OTEL, so providing a `daemon.json` option would only be shifting the goalposts.

### Defaults (enabled by default, disabled by default)

We discussed resource-consumption, defaults, and options to enable/disable OTEL;

- OTEL works asynchronously (uses goroutines), so won't _directly_ affect performance of code-paths, but it _may_ take some resources (mostly the exporting).
- While OTEL can be very useful, there may be cases where resources are constrained, and we do want to provide an option to fully disable OTEL for such situations (regardless any OTEL config on the host).
- Configuring OTEL (as described earlier) is handled through environment variables. Setting those variables requires access to the host, which we consider to be "secure" (if you control the host, it's reasonable to control this setting for the daemon).
- _However_, the OTEL library also has a fallback to connect to `localhost:4318`. From the [`WithEndpoint()` docs](https://pkg.go.dev/go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetrichttp@v0.39.0#WithEndpoint):

    > If the OTEL_EXPORTER_OTLP_ENDPOINT or OTEL_EXPORTER_OTLP_METRICS_ENDPOINT environment variable is set, and this option is not passed, that variable value will be used. If both are set, OTEL_EXPORTER_OTLP_METRICS_ENDPOINT will take precedence.
    >
    > By default, if an environment variable is not set, and this option is not passed, "localhost:4318" will be used.

We discussed this fallback;

- The fallback can happan at any moment (i.e., it's not just at startup); OTEL will continue to send to `localhost:4318` if no endpoint is configured (through env-vars).
- Post `4318` is a non-privileged port (not in the "0-1024" port range)

We found this fallback to be non-desirable for the default configuration, because:

- The user did not explicitly opt-in to use OTEL (i.e., no environment variables were set up), and the user may not be aware of this happening.
- :warning: OTEL continuing to try `localhost;4318` means that (for example, on a "shared host"), a container may be started that listens on port `4318` and now starts collecting telemetry (which could potentially be of "any system" on the host that supports OTEL).
- Of course, running "untrusted" containers on a machine is always considered unsafe, but we also consider that not all users may consider this.
- While the data sent should generally not contain sensitive data (e.g. no passwords, env-vars content), it may leak information, such as the names of images being pulled. It's also possible for logs to be sent (e.g., see https://github.com/containerd/containerd/pull/6003).

Based on the above, we decided that we want to provide a daemon option to control this behavior, and to make the `localhost` fallback opt-in. This would look something like (name of the option and option-values TBD);


| Flag          | Value              | Description                                                                                                                       |
|---------------|--------------------|-----------------------------------------------------------------------------------------------------------------------------------|
| `--telemetry` | disabled           | Disable OTEL.                                                                                                                     | 
| `--telemetry` | from-env (default) | Enable OTEL if the OTEL environment variables are set, and otherwise disable OTEL.                                                | 
| `--telemetry` | auto               | "The fully monty". Configure through environment variables (if set), and otherwise continue trying connecting to `localhost:4318` |
| `--telemetry` | other values       | ERROR: invalid value                                                                                                              |


There may be some challenges to implement this;

- The OTEL library uses the fallback by default, and (to our knowledge) does not have an option to disable the fallback (other than, e.g., setting `OTEL_EXPORTER_OTLP_ENDPOINT` to an invalid address). 
- The list of available env-vars is still expanding, which means that it may not be possible for us to check whether or not OTEL should be enabled (based on environment variables).
    - BuildKit _does_ implement this logic already, and implemented a utility for this [see `util/tracing/detect/otlp.otlpExporter()`](https://github.com/moby/buildkit/blob/2ff0d2a2f53663aae917980fa27eada7950ff69c/util/tracing/detect/otlp.go#L18C1-L45), but it not being part of the OTEL library poses the risk that it gets out of sync with the library. We could consider contributing it to upstream though.
    - While discussing that code, we also brought up that the OTEL library should provide a utility to construct the right client, i.e. [this part of the code](https://github.com/moby/buildkit/blob/2ff0d2a2f53663aae917980fa27eada7950ff69c/util/tracing/detect/otlp.go#L34-L42) should not be left to consumers of the library, and the OTEL library should provide a utility for this (perhaps configurable with a allow/deny list of options?)
    - :info: While looking in upstream, we _did_ find an autoexporter "contrib" package in https://github.com/open-telemetry/opentelemetry-go-contrib/blob/9a582bd098a287f12e107e17d884a124d54606ab/exporters/autoexport/registry.go#L131-L147, which _may_ be what we're looking for?


> #### :warning: containerd's defaults
>
> While (of course) other systems supporting OTEL still sending to `localhost:4318` is out of scope for this project, we did discuss containerd. While discussing this topic, @cpuguy83 mentioned that current releases of containerd (1.7 and before) do not use the `localhost:4318` fallback, but the `main` branch (v2.0.0-dev) _does_. He mentioned that there was some discussion about this (can't find it though), but at the time it was not considered an issue. Following the discussion we had on this project, he is considering to revive the discussion on the containerd repository (and to propose changing the default).

### Future expansion

(n.b.: this part was not discussed during the call)

The `--telemetry` (name TBH) flag takes a "string" value, which leaves room for future expansion if we see a need, and which can be done without breaking backward compatibility;

Additional values can be added to the list, e.g. an option to "only look at localhost";

```bash
dockerd --telemetry=localhost
```

We can introduce an advanced (CSV) syntax;

```bash
# equivalent of:
#
# dockerd --telemetry=auto
dockerd --telemetry config=auto

# equivalent of:
#
#     OTEL_TRACES_EXPORTER=otlp OTEL_EXPORTER_OTLP_ENDPOINT="..."  "OTEL_EXPORTER_OTLP_TRACES_ENDPOINT": "..." dockerd
#
dockerd --telemetry exporter=otlp,endpoint="...",traces-endpoint="..."
```

Or, alternatively, we can add a `--telemetry-opt` option, or options for each individual config (`--telemetry-traces-exporter=...`).

cpuguy83 commented:
@thaJeztah PTAL at https://github.com/cpuguy83/docker/blob/4196286d6a3750f7c01a6c23db5984683bbbbaff/cmd/dockerd/tracing.go#L19-L128

This is utilizing the standard otel env vars.

thaJeztah commented:
Adding some quick notes to my earlier comment (https://github.com/moby/moby/pull/45652#issuecomment-1619805977), after discussing in another maintainers call;

OTEL Env-vars are documented here;
https://opentelemetry.io/docs/specs/otel/configuration/sdk-environment-variables/

It was pointed out that there's a `OTEL_SDK_DISABLED` env-var:

> Boolean value. If “true”, a no-op SDK implementation will be used for all telemetry signals. Any other value or absence of the variable will have no effect and the SDK will remain enabled. This setting has no effect on propagators configured through the OTEL_PROPAGATORS variable.

Combined with the logic (similar to BuildKit) to only enable if OTEL is configured (so no default to look for `localhost:4318` by default)
https://github.com/moby/buildkit/blob/2ff0d2a2f53663aae917980fa27eada7950ff69c/util/tracing/detect/otlp.go#L18C1-L45

We may forego the daemon.json and command-line options (for now)

cpuguy83 committed:
Add otel support

This uses otel standard environment variables to configure tracing in
the daemon.
It also adds support for propagating trace contexts in the client and
reading those from the API server.

See
https://opentelemetry.io/docs/specs/otel/configuration/sdk-environment-variables/
for details on otel environment variables.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

cpuguy83 committed:
Wire up tests to support otel tracing

Integration tests will now configure clients to propagate traces as well
as create spans for all tests.

Some extra changes were needed (or desired for trace propagation) in the
test helpers to pass through tracing spans via context.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

cpuguy83 committed:
CI: Setup otel tracing for integration tests

This wires up the integration tests to export spans to a jager instance.
After tests are finished it exports the data out of jaeger and uploads
as an artifact to the action run.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

cpuguy83 commented:
This is ready for review.
I recommend you review this by commit.
That said, there is one giant commit with all the changes needed to wire through tracing into integration{-cli} tests.... so yeah good luck loading that.

thaJeztah commented on client/client.go:
```diff
@@ -198,6 +202,15 @@ func NewClientWithOpts(ops ...Opt) (*Client, error) {
 	}
 
+	// TODO: Maybe filter the _ping endpoint from otel tracing?
+	client.Transport = otelhttp.NewTransport(
```
Will the extra wrapping affect things like `client.Close()` https://github.com/moby/moby/blob/96b473a0bdc3f4c1cd0324f3ff044b04ab349df8/client/client.go#L218-L221


And `client.Dialer()` ? https://github.com/moby/moby/blob/96b473a0bdc3f4c1cd0324f3ff044b04ab349df8/client/client.go#L352

cpuguy83 replied:
Good catch... we need to fix those.

cpuguy83 replied:
Fixed.

thaJeztah commented on cmd/dockerd/daemon.go:
```diff
@@ -228,7 +232,26 @@ func (cli *DaemonCli) start(opts *daemonOptions) (err error) {
 	preNotifyReady()
 
-	sp := setupTracing(ctx)
+	// Setup buildkit trace recorder
+	detect.Recorder = detect.NewTraceRecorder()
```
Silly question; is there a specific reason to put this code in this part of the start-up cycle? Would there be benefits in setting things up as early as possible to trace as much of the flow as possible?

cpuguy83 replied:
No I don't think it would change anything.
This is for tracing builds which wouldn't have anything running at this point.

cpuguy83 replied:
Actually, I should likely move this below our tracing config... I'll have to double check what's happening in this function.

cpuguy83 replied:
Where it is is good.
All its doing:

```go
func NewTraceRecorder() *TraceRecorder {
	tr := &TraceRecorder{
		m:         map[trace.TraceID]*stubs{},
		listeners: map[trace.TraceID]int{},
	}

	go func() {
		t := time.NewTimer(60 * time.Second)
		for {
			<-t.C
			tr.gc()
			t.Reset(50 * time.Second)
		}
	}()

	return tr
}
```

thaJeztah commented on cmd/dockerd/daemon.go:
```diff
@@ -335,8 +358,8 @@ func (cli *DaemonCli) start(opts *daemonOptions) (err error) {
 
-	if sp != nil {
-		sp.Shutdown(context.Background())
+	if tp != nil {
```
LOL, I went looking for an `sp` variable in the code (as I was looking per-commit). I guess the original `sp` was a typo for `tp` ?

cpuguy83 replied:
`span processor`
I changed what got returned.

thaJeztah replied:
Ah! `sp` makes sense now 👍

thaJeztah commented on cmd/dockerd/tracing.go:
```diff
@@ -82,13 +84,15 @@ func setupTracing(ctx context.Context) sdktrace.SpanProcessor {
+	}
+
+	if _, ok := os.LookupEnv(otelServiceNameEnv); !ok {
+		os.Setenv(otelServiceNameEnv, filepath.Base(os.Args[0]))
```
Is this the "identifier" for the service? If so, would it make sense to use the daemon's ID for this? (instead of the file-path)?

cpuguy83 replied:
Yes, that could work. This is what I do for the integration test daemons, actually (by setting the env var).

cpuguy83 replied:
Updated to include the daemon ID.

thaJeztah commented on integration-cli/requirements_unix_test.go:
```diff
@@ -11,9 +11,12 @@ import (
 )
 
-// SysInfo stores information about which features a kernel supports.
 var SysInfo *sysinfo.SysInfo
```
Any reason why you removed this comment? Or was the intent to un-export it?

cpuguy83 replied:
Yeah, probably intended to unexport this.

thaJeztah commented on integration/config/config_test.go:
```diff
@@ -76,39 +76,46 @@ func TestConfigList(t *testing.T) {
 		filters  filters.Args
 		expected []string
 	}{
-		// test filter by name `config ls --filter name=xxx`
 		{
+			desc:     "test filter by name",
 			filters:  filters.NewArgs(filters.Arg("name", testName0)),
```
❤️ 

Guess we could already get these in separately, but not critical, so I'm ok with keeping it as part of this PR

thaJeztah commented on integration-cli/check_test.go:
```diff
@@ -38,37 +42,64 @@ const (
+		shutdown(ctx)
+	}()
+
+	baseContext = ctx
```
I wish `testing.T` / `testing.M` would have the ability to attach a context (so that we didn't need to use a package-level var for this) 😢

thaJeztah commented on integration/network/helpers.go:
```diff
@@ -9,42 +9,43 @@ import (
+func CreateMasterDummy(ctx context.Context, t *testing.T, master string) {
 	// ip link add <dummy_name> type dummy
-	icmd.RunCommand("ip", "link", "add", master, "type", "dummy").Assert(t, icmd.Success)
-	icmd.RunCommand("ip", "link", "set", master, "up").Assert(t, icmd.Success)
+	testutil.RunCommand(ctx, "ip", "link", "add", master, "type", "dummy").Assert(t, icmd.Success)
+	testutil.RunCommand(ctx, "ip", "link", "set", master, "up").Assert(t, icmd.Success)
```
I see we're replacing `icmd` for our wrapper here, but we don't do that everywhere (e.g., we didn't do it in the `integration-cli`.

- Is that something we want to do across the board? (if so, perhaps we should create a tracking ticket for that)
- Also wondering if passing a context is something that would make sense for `gotest.tools/v3/icmd"` (which we could contribute) 🤔  - probably for a `v4` then 😂

thaJeztah reviewed (commented):
Thanks! This looks really great!


I went through all commits, and went through all files locally for the "large commit".


I left some comments on the first few commits, and I think I found a missing `ctx :=` in one of the commits (due to the later test-fixes for the daemon tests).

I also gave it a quick go to move the last 3 commits to a separate PR, as those fixes should (ideally) already be fixed on master (and possibly backported);

- https://github.com/moby/moby/pull/46122


I tried a quick rebase of your PR branch on top of that (to see how bad it is), but it was not too complicated (but I'm happy to push my rebase if it helps you).

cpuguy83 commented on integration/daemon/daemon_test.go:
```diff
@@ -178,27 +177,34 @@ func TestDaemonProxy(t *testing.T) {
+			"NO_PROXY=example.com",
+		))
 		c := d.NewClientT(t)
-		defer func() { _ = c.Close() }()
```
It was removed in the TestDaemonProxy changes because I moved the close `NewClientT` as a `t.Cleanup` 😆

thaJeztah replied:
Ah! I should've spotted that, thanks!

cpuguy83 commented on client/client.go:
```diff
@@ -200,6 +215,14 @@ func NewClientWithOpts(ops ...Opt) (*Client, error) {
 	}
 
+	c.client.Transport = otelhttp.NewTransport(
+		c.client.Transport,
```
Yeah its an issue for its not an `*http.Transport`

cpuguy83 commented on vendor.mod:
```diff
@@ -105,6 +109,8 @@ require (
 	resenje.org/singleflight v0.3.0
 )
 
+require go.opentelemetry.io/otel/exporters/otlp/otlptrace v1.4.1
+
```
I recall trying to move this and it put it right back.

thaJeztah replied:
Yeah, it can have a mind of its own at times 😞 ... it _should_ work though 🤞

cpuguy83 replied:
Looks like it worked this time.

cpuguy83 commented on integration/daemon/daemon_test.go:
```diff
@@ -328,33 +329,26 @@ func TestDaemonProxy(t *testing.T) {
 	t.Run("reload sanitized", func(t *testing.T) {
+		t.Parallel()
+
 		_ = testutil.StartSpan(ctx, t)
```
That's in a later commit. I should squash those down.

cpuguy83 commented on client/hijack.go:
```diff
@@ -59,10 +64,31 @@ func fallbackDial(proto, addr string, tlsConfig *tls.Config) (net.Conn, error) {
+		}
+	}
+
+	ctx, span := tp.Tracer("").Start(ctx, req.Method+" "+req.URL.Path)
```
I could add a comment, though I suspect after this we may want to add more custom spans where we'd be doing the same thing.
To me the documentation is in the interface.

thaJeztah replied:
Oh! question on the same line; as we're shadowing `ctx`  here, would it be clearer to;

```suggestion
	var span trace.Span
	ctx, span = tp.Tracer("").Start(ctx, req.Method+" "+req.URL.Path)
```

cpuguy83 replied:
Why?

thaJeztah replied:
Ah, sorry, yeah currently it doesn't add much. I ended up in a rebase / branch where this line ended up in a "conditional" (if), in which case the "shadowing" became an actual "shadow". Hence my thinking if it would be better to be defensive here. But agreed, that it's not super-important.

cpuguy83 commented on Makefile:
```diff
@@ -75,7 +75,9 @@ DOCKER_ENVS := \
 	-e PRODUCT \
-	-e PACKAGER_NAME
+	-e PACKAGER_NAME \
+	-e OTEL_EXPORTER_OTLP_ENDPOINT \
+	-e OTEL_SERVICE_NAME
```
> what if OTEL_EXPORTER_OTLP_ENDPOINT is something like localhost:12345

It will fail to connect and show error messages in the daemon/client logs, but wouldn't block anything, except maybe shutting down the tracer would error.

Basically, don't do that.

> do we need to have something to allow passing a socket (and bind-mount that into the container?)

I don't think there is any mechanism to connect over anything but TCP.

cpuguy83 commented:
Requested changes applied and all green (or I suspect all green, I think we hit a flakey libnetwork test in jenkins).

tonistiigi commented on cmd/dockerd/daemon.go:
```diff
@@ -228,6 +232,35 @@ func (cli *DaemonCli) start(opts *daemonOptions) (err error) {
 	preNotifyReady()
 
+	// Setup buildkit trace recorder
+	detect.Recorder = detect.NewTraceRecorder()
```
Not sure this is the best place to put this (to me it can be left in the init as it is just a variable definition and no slow path).

Afaics, there are now 2 separate trace providers, one coming from `getTracerProvider()` and the other from `detect.TraceProvider()`. The second one will handle only the requests coming to the `/grpc` and only one that buildkit sees. This may be fine, but in that case it makes little sense to initialize `detect.Recorder` here together with `getTracerProvider` because they don't have any shared parts. Technically I think the current code works as the first code calling `detect.TraceProvider()` is in `NewRouter`, so called after this codepath.

cpuguy83 replied:
I changed the grpc router to just read the tracer provider from the otel lib.
I definitely prefer to have the recorder set here where we control ordering rather than in an init.

tonistiigi replied:
I don't understand the latest changes. If the requests coming in from the grpc interceptors use a different trace provider this will mean that the spans will not appear in buildkit traces as well as that initialization of `Recorder` does not do anything.

My comment about the location on recorder initialization was that it's not related to `getTracerProvider()` and although the current naming seems to suggest it.

cpuguy83 replied:
Ok, I'm stepping through buildkit to see how this is used and see the problem now.
`detect.TracerProvider()` is using a new tracer provider to wrap any configured span exporter with the recorder.

How can we test this so it doesn't break?

tonistiigi replied:
In buildctl you can do:
```
buildctl debug histories --format '{{json .}}'
// find a build and a trace record
buildctl debug get sha256:db7dbd0f7d0d2239075f83ec6b2dd76bd7d9558bdb79895242c7a3944c6dadaf
```

If you have Docker Desktop installed then alternative:
```
curl --unix-socket ~/Library/Containers/com.docker.docker/Data/docker-desktop-build.sock http://localhost/builds/<buildx builder>/<buildx node>/<build ref>/stream
// get trace
curl --unix-socket ~/Library/Containers/com.docker.docker/Data/docker-desktop-build.sock 'http://localhost/blobs/<buildx builder>/<buildx node>?digest=sha256:8727e058fe0ed45dac529e8c8390107a001e545d30a83c6c148a2aa8d5058df9'
```

cpuguy83 replied:
Ok, I've added a test case for this.
The reason this was taking me so long is not realizing the trace data gets set on the history record async (which of course makes sense) so it looked like it wasn't populating.
Validated the test fails when this was setup incorrectly.

Also validated buildkit client and server traces are exported with the rest of the daemon traces.
<img width="858" alt="image" src="https://github.com/moby/moby/assets/799078/e9ee20a4-b730-4213-9693-6bcc46e1918b">

thaJeztah commented:
@cpuguy83 I pushed a branch that rebases this PR on top of my separate "fix TestDaemonProxy integration tests" PR;

- https://github.com/moby/moby/pull/46126

Let me know if you're ok with merging that one (https://github.com/moby/moby/pull/46126) first

thaJeztah reviewed (approved):
LGTM

neersighted commented on .github/actions/setup-tracing/action.yml:
```diff
@@ -0,0 +1,12 @@
+name: 'Setup Tracing'
+description: 'Composite action to set up the tracing jor test jobs'
```
```suggestion
description: 'Composite action to set up the tracing for test jobs'
```

cpuguy83 replied:
lol.
Fixed.

thaJeztah commented:
Missing some import somewhere, @cpuguy83 

```
integration/container/pidmode_linux_test.go:41:8: invalid operation: cannot call non-function setupTest(t) (value of type context.Context)
integration/container/pidmode_linux_test.go:43:9: undefined: context

thaJeztah commented:
😢 and now vendor validation is failing; looks like an "requires" is in the wrong section of the `vendor.mod`;

```patch
diff --git a/vendor.mod b/vendor.mod
index c16d3e2e0a..f93e4d862a 100644
--- a/vendor.mod
+++ b/vendor.mod
@@ -93,7 +93,7 @@ require (
 	go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelg
rpc v0.29.0
 	go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp v0.29.0
 	go.opentelemetry.io/otel v1.4.1
-	go.opentelemetry.io/otel/exporters/otlp/otlptrace v1.4.1
+	go.opentelemetry.io/otel/exporters/otlp/otlptrace v1.4.1 // indirect

 	go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc v1.4.1
 	go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp v1.4.1
 	go.opentelemetry.io/otel/sdk v1.4.1
```


(don't shoot the messenger)

thaJeztah commented:
oh! looks like a test-failure as well (on Jenkins);

```
=== RUN   TestBuildkitHistoryTracePropagation
time="2023-08-24T22:40:31Z" level=info msg="[core] [Channel #1] Channel created" module=grpc
time="2023-08-24T22:40:31Z" level=info msg="[core] [Channel #1] original dial target is: \"unix:///run/buildkit/buildkitd.sock\"" module=grpc
time="2023-08-24T22:40:31Z" level=info msg="[core] [Channel #1] parsed dial target is: {Scheme:unix Authority: URL:{Scheme:unix Opaque: User: Host: Path:/run/buildkit/buildkitd.sock RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}}" module=grpc
time="2023-08-24T22:40:31Z" level=info msg="[core] [Channel #1] Channel authority set to \"localhost\"" module=grpc
time="2023-08-24T22:40:31Z" level=info msg="[core] [Channel #1] Resolver state updated: {\n  \"Addresses\": [\n    {\n      \"Addr\": \"/run/buildkit/buildkitd.sock\",\n      \"ServerName\": \"\",\n      \"Attributes\": {},\n      \"BalancerAttributes\": null,\n      \"Type\": 0,\n      \"Metadata\": null\n    }\n  ],\n  \"ServiceConfig\": null,\n  \"Attributes\": null\n} (resolver returned new addresses)" module=grpc
time="2023-08-24T22:40:31Z" level=info msg="[core] [Channel #1] Channel switches to new LB policy \"pick_first\"" module=grpc
time="2023-08-24T22:40:31Z" level=info msg="[core] [Channel #1 SubChannel #2] Subchannel created" module=grpc
time="2023-08-24T22:40:31Z" level=info msg="[core] [Channel #1] Channel Connectivity change to CONNECTING" module=grpc
time="2023-08-24T22:40:31Z" level=info msg="[core] [Channel #1 SubChannel #2] Subchannel Connectivity change to CONNECTING" module=grpc
time="2023-08-24T22:40:31Z" level=info msg="[core] [Channel #1 SubChannel #2] Subchannel picks a new address \"/run/buildkit/buildkitd.sock\" to connect" module=grpc
time="2023-08-24T22:40:31Z" level=info msg="[core] [Channel #1 SubChannel #2] Subchannel Connectivity change to READY" module=grpc
time="2023-08-24T22:40:31Z" level=info msg="[core] [Channel #1] Channel Connectivity change to READY" module=grpc
time="2023-08-24T22:40:31Z" level=info msg="[core] [Server #4] Server created" module=grpc
    build_traces_test.go:83: Waiting for trace to be available
    build_traces_test.go:84: timeout hit after 10s: trace not available yet
time="2023-08-24T22:40:41Z" level=info msg="[core] [Channel #1] Channel Connectivity change to SHUTDOWN" module=grpc
time="2023-08-24T22:40:41Z" level=info msg="[core] [Channel #1 SubChannel #2] Subchannel Connectivity change to SHUTDOWN" module=grpc
time="2023-08-24T22:40:41Z" level=info msg="[core] [Channel #1 SubChannel #2] Subchannel deleted" module=grpc
time="2023-08-24T22:40:41Z" level=info msg="[core] [Channel #1] Channel deleted" module=grpc
--- FAIL: TestBuildkitHistoryTracePropagation (10.09s)
```

cpuguy83 commented:
I knew that was going to happen at some point when tracing isn't setup on a test env.
Updating the test now.

cpuguy83 committed:
Add test for buildkit history trace propagation

This test ensures that we are able to propagate traces into buildkit's
history API.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

tonistiigi reviewed (commented):
Tested this and it now logs "no trace recorder found, skipping" for every build. It happens because `detect.Exporter()` returns nil.

(This can be a follow-up as not directly related, but) I don't think client-side traces can work unless `TraceCollector` is set in `control.NewController(control.Opt{`

cpuguy83 commented on cmd/dockerd/tracing.go:
```diff
@@ -0,0 +1,156 @@
+
+	var sampler sdktrace.Sampler
+
+	samplerValue := getEnv(otelTracesSamplerEnv)
```
Oh hmmm... I don't know why I was thinking we always had to define a sampler. Maybe just from older otel implementations or looking at the original containerd implementation I thinking it was required.
I'll take a look.

cpuguy83 replied:
Where do you see this setup?
What I'm seeing is it only has the default when a sampler is not provided and does not read env vars.

In NewTracerProvider: 

https://github.com/open-telemetry/opentelemetry-go/blob/3c476ce1816ae6f38758e90cc36d8b77ebcc223b/sdk/trace/provider.go#L112

Which is calling:

https://github.com/open-telemetry/opentelemetry-go/blob/3c476ce1816ae6f38758e90cc36d8b77ebcc223b/sdk/trace/provider.go#L489-L500

tonistiigi replied:
https://github.com/open-telemetry/opentelemetry-go/blob/3c476ce1816ae6f38758e90cc36d8b77ebcc223b/sdk/trace/provider.go#L106

https://github.com/open-telemetry/opentelemetry-go/blob/3c476ce1816ae6f38758e90cc36d8b77ebcc223b/sdk/trace/provider.go#L476

https://github.com/open-telemetry/opentelemetry-go/blob/3c476ce1816ae6f38758e90cc36d8b77ebcc223b/sdk/trace/sampler_env.go#L61

cpuguy83 commented on cmd/dockerd/tracing.go:
```diff
@@ -0,0 +1,156 @@
+var errTracingDisabled = errors.New("tracing disabled")
+
+func getTracerProvider(ctx context.Context, getEnv func(string) string) (*sdktrace.TracerProvider, error) {
+	// By default the OTLP libs will connect to localhost if no endpoint is set.
```
After much head->desk:

Turns out, all the difference is this code was defaulting to http/prorobuf for OTLP, detect defaults to grpc.
This is why I wasn't getting traces in my exporter.
Per spec, default should be http/protobuf: https://opentelemetry.io/docs/specs/otel/protocol/exporter/#specify-protocol

I don't want to break existing users of buildkit traces... but also it seems more correct to go with the spec for tracing the daemon.... anyone have any thoughts on this?

cpuguy83 replied:
Spec:

> If no configuration is provided the default transport SHOULD be http/protobuf unless SDKs have good reasons to choose grpc as the default (e.g. for backward compatibility reasons when grpc was already the default in a stable SDK release).

tonistiigi replied:
I don't remember where I took this code anymore. I thought it might have something to do with UNIX and HTTP not working well together for the socket we send to containers, but that code sets a specific protocol. https://github.com/moby/buildkit/blob/e2dfabdebce24c9f55623259160a6a9faa1cb31e/executor/oci/spec.go#L44

Assuming the only thing it can break is external OTLP collectors that don't follow spec, I think we should just change it in buildkit. However, it wouldn't be nice to change it in a patch release.

tonistiigi replied:
When I wrote the `detect` package, the (implicit) default was still grpc.
https://github.com/open-telemetry/opentelemetry-specification/pull/1969/commits/0297525a5634e233533bc5b14b15b69fae18abb8
https://github.com/moby/buildkit/commit/292f1faf1209bb2a49fa4a3e9c3f785c51afbe55#diff-77fb71b07405102c7882eae6d80aa8300966cb29ffe2d478a63b46899602f1f8R29

cpuguy83 replied:
Yep, definitely not surprised the default changed 😸 

I've updated this PR to use detect for everything and also set the otlp protocol to http/profobuf if unset so that `detect` will pick it up.
LMK if this sounds good.

tonistiigi commented on builder/builder-next/controller.go:
```diff
@@ -136,11 +147,12 @@ func newSnapshotterController(ctx context.Context, rt http.RoundTripper, opt Opt
+		HistoryConfig:  historyConf,
+		LeaseManager:   wo.LeaseManager,
+		ContentStore:   wo.ContentStore,
+		TraceCollector: getTraceExporter(ctx),
```
Note that the client trace delegation still doesn't work yet, but the issue is in buildx side where it is not enabled for all drivers. I'll make a patch for this tomorrow. Moby side looks correct.

thaJeztah commented on cmd/dockerd/daemon.go:
```diff
@@ -227,6 +232,28 @@ func (cli *DaemonCli) start(opts *daemonOptions) (err error) {
 
+	const otelServiceNameEnv = "OTEL_SERVICE_NAME"
+	if _, ok := os.LookupEnv(otelServiceNameEnv); !ok {
+		id, err := daemon.LoadOrCreateID(cli.Config.Root)
```
Yeah, that's kinda ugly. So I _think_ the motivation here may have been because @cpuguy83 ran OTEL on the integration tests, in which case there's "many" instances of `dockerd` (and sometimes multiple used in the same test).

Other cases could also mean if you collect OTEL centrally, but collect it for multiple hosts; I suggested to use the daemon "ID" for this, because that's an ID that's already visible to the user (part of `docker info` output).

But I agree that it's just "noise" if you only have a single daemon.

I think this would be fine to tweak in a follow-up, but we need to think what the best approach would be;

- default to `dockerd` (no ID after)
- perhaps allow `OTEL_SERVICE_NAME` to take a "template" (`dockerd-{{.ID}}`, `dockerd-{{.HostName}}`, etc..) and we could fill in that template.

🤔

cpuguy83 replied:
https://github.com/moby/moby/pull/45652#discussion_r1272545429 😄 

I think what would work is adding a span id to the daemon logs (when available) like bklog does would work to differentiate daemons.

thaJeztah reviewed (approved):
had a discussion in the maintainers call;

- the integration tests already set the SERVICE_NAME (so for our own tests, it's not an issue to remove the daemon iD)
- the current code moved "generating" the daemon-id only for OTEL (not ideal)
- let's remove the ID for now, and work on other options in a follow-up

after that, "LGTM" (let's get this in!)

cpuguy83 commented:
Updated and all is green.

thaJeztah commented:
w000000000t! thank so much for this, @cpuguy83 !

neersighted commented:
@cpuguy83 I assume you'll be opening the GHA followup PRs?

thaJeztah commented:
Now that this one's merged, I rebased, moved this one out of draft 😅 (wink wink, nudge nudge, say no more)

- https://github.com/moby/moby/pull/46088

cpuguy83 commented:
> @cpuguy83 I assume you'll be opening the GHA followup PRs?

https://github.com/moby/moby/pull/46429
